### ISSUE TITLE
[no-unnecessary-condition] False unnecessary conditional with nullish coalescing

### ISSUE THREAD

**Issue by whocaresk** [typescript-eslint/typescript-eslint#3750]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": ["error"]
  }
}
```

```TS
let test: any[] | undefined = undefined;

/* some code where test var will be assigned (or not) */

const tested = test ?? [];
```

**Expected Result**
No ESLint errors reported.

**Actual Result**
``ESLint: Unnecessary conditional, expected left-hand side of `??` operator to be possibly null or undefined.(@typescript-eslint/no-unnecessary-condition)``

**Additional Info**  
I'm using the strictNullChecks compiler flag.    
Everything is fine until there is appears an array notation. Possibly also doesn't work with any other type params with generics. Is that even supposed to work?

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.29.1` |
| `@typescript-eslint/parser`        | `4.29.1` |
| `TypeScript`                       | `4.3.5` |
| `ESLint`                           | `7.32.0` |
| `node`                             | `14.17.3` |

**Comment by whocaresk** [comment]
There is more that may be related:   
```typescript
const getProp = (val: any): string[] | undefined => val?.key;
const propValue = getProp({});

if (!propValue) {
  return undefined;
}
```
outputs: ``ESLint: Unnecessary conditional, value is always falsy.(@typescript-eslint/no-unnecessary-condition)``.  
If i remove array mark from string - then everything is fine.

**Comment by bradzacher** [comment]
I'm unable to repro this against master.

![image](https://user-images.githubusercontent.com/7462525/129943999-745ef075-2081-49a9-bd74-0690dd06cea0.png)

Please create an isolated repro repo.

**Comment by whocaresk** [comment]
After full reboot problem is gone.   
Don't know why restarting IDE either restarting language service doesn't help me yesterday.
